File: runtime/cpp/titan-ast-runtime-lib/PersistentData.cpp
//
// Created by tian wei jun on 2022/11/24 0024.
//

#include "PersistentData.h"
#include "SyntaxDfaState.h"

PersistentData::PersistentData()
    : intByteBuffer(ByteBuffer(4, true)), stringPool(nullptr),
      sizeOfStringPool(0), grammars(nullptr), sizeOfGramamrs(0),
      productionRules(nullptr), sizeOfProductionRules(0) {}

BuildAutomataResult PersistentData::init(const std::string *automataFilePath) {
  inputStream.open(*automataFilePath, std::ios::in | std::ios::binary);

  if (!inputStream.is_open()) {
    return {false, "open automata File error,path:'" + *automataFilePath + "'"};
  }
  return {true, ""};
}

PersistentData::~PersistentData() {
  inputStream.close();
  // other delete by AutomataData
}

GetProductionRulesByInputStreamResult
PersistentData::getProductionRulesByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetProductionRulesByInputStreamResult{false};
  }
  int countOfProductionRules = readIntResult.data;
  auto **heapProductionRules = new ProductionRule *[countOfProductionRules];
  for (int indexOfProductionRule = 0;
       indexOfProductionRule < countOfProductionRules;
       indexOfProductionRule++) {
    heapProductionRules[indexOfProductionRule] = new ProductionRule();
  }

  this->sizeOfProductionRules = countOfProductionRules;
  this->productionRules = heapProductionRules;

  for (int indexOfProductionRule = 0;
       indexOfProductionRule < countOfProductionRules;
       indexOfProductionRule++) {
    ProductionRule *productionRule = heapProductionRules[indexOfProductionRule];
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteProductionRules(productionRules, sizeOfProductionRules);
      return GetProductionRulesByInputStreamResult{false};
    }
    productionRule->grammar = grammars[readIntResult.data];
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteProductionRules(productionRules, sizeOfProductionRules);
      return GetProductionRulesByInputStreamResult{false};
    }
    int indexOfAliasInStringPool = readIntResult.data;
    if (indexOfAliasInStringPool >= 0) {
      productionRule->alias = stringPool[indexOfAliasInStringPool];
    }
    GetSyntaxDfaByInputStreamResult getSyntaxDfaByInputStreamResult =
        getSyntaxDfaByInputStream();
    if (!getSyntaxDfaByInputStreamResult.isOk) {
      deleteProductionRules(productionRules, sizeOfProductionRules);
      return GetProductionRulesByInputStreamResult{false};
    }
    productionRule->reducingDfa = getSyntaxDfaByInputStreamResult.data;
  }
  return GetProductionRulesByInputStreamResult{true, heapProductionRules,
                                               countOfProductionRules};
}

void PersistentData::deleteProductionRules(ProductionRule **productionRules,
                                           int sizeOfProductionRules) {
  for (int i = 0; i < sizeOfProductionRules; i++) {
    delete productionRules[i];
  }
  delete productionRules;
}

GetSyntaxDfaByInputStreamResult PersistentData::getSyntaxDfaByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetSyntaxDfaByInputStreamResult{false, nullptr};
  }
  int sizeOfSyntaxDfaStates = readIntResult.data;
  auto **syntaxDfaStates = new SyntaxDfaState *[sizeOfSyntaxDfaStates];
  for (int indexOfSyntaxDfaState = 0;
       indexOfSyntaxDfaState < sizeOfSyntaxDfaStates; indexOfSyntaxDfaState++) {
    syntaxDfaStates[indexOfSyntaxDfaState] =
        new SyntaxDfaState(indexOfSyntaxDfaState);
  }
  // countOfSyntaxDfaStates-(type-countOfEdges-[ch,dest]{countOfEdges}-countOfProductions-productions)
  for (int indexOfSyntaxDfaState = 0;
       indexOfSyntaxDfaState < sizeOfSyntaxDfaStates; indexOfSyntaxDfaState++) {
    SyntaxDfaState *syntaxDfaState = syntaxDfaStates[indexOfSyntaxDfaState];
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteSyntaxDfaStates(syntaxDfaStates, sizeOfSyntaxDfaStates);
      return GetSyntaxDfaByInputStreamResult{false, nullptr};
    }
    syntaxDfaState->type = readIntResult.data;
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteSyntaxDfaStates(syntaxDfaStates, sizeOfSyntaxDfaStates);
      return GetSyntaxDfaByInputStreamResult{false, nullptr};
    }
    int sizeOfEdges = readIntResult.data;
    syntaxDfaState->edges =
        std::unordered_map<const Grammar *, SyntaxDfaState *,
                           PtrGrammarContentHash, PtrGrammarContentEq>(
            sizeOfEdges);
    for (int indexOfEdge = 0; indexOfEdge < sizeOfEdges; indexOfEdge++) {
      readIntResult = readInt();
      if (!readIntResult.isOk) {
        deleteSyntaxDfaStates(syntaxDfaStates, sizeOfSyntaxDfaStates);
        return GetSyntaxDfaByInputStreamResult{false, nullptr};
      }
      Grammar *ch = grammars[readIntResult.data];
      readIntResult = readInt();
      if (!readIntResult.isOk) {
        deleteSyntaxDfaStates(syntaxDfaStates, sizeOfSyntaxDfaStates);
        return GetSyntaxDfaByInputStreamResult{false, nullptr};
      }
      SyntaxDfaState *chToState = syntaxDfaStates[readIntResult.data];
      std::pair<const Grammar *, SyntaxDfaState *> keyValue(ch, chToState);
      syntaxDfaState->edges.insert(keyValue);
    }
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteSyntaxDfaStates(syntaxDfaStates, sizeOfSyntaxDfaStates);
      return GetSyntaxDfaByInputStreamResult{false, nullptr};
    }
    int sizeOfProductions = readIntResult.data;
    for (int indexOfProduction = 0; indexOfProduction < sizeOfProductions;
         indexOfProduction++) {
      readIntResult = readInt();
      if (!readIntResult.isOk) {
        deleteSyntaxDfaStates(syntaxDfaStates, sizeOfSyntaxDfaStates);
        return GetSyntaxDfaByInputStreamResult{false, nullptr};
      }
      syntaxDfaState->closingProductionRules.push_back(
          productionRules[readIntResult.data]);
    }
    syntaxDfaState->closingProductionRules.shrink_to_fit();
  }
  return GetSyntaxDfaByInputStreamResult{
      true, new SyntaxDfa(syntaxDfaStates[0],
                          (const SyntaxDfaState **)syntaxDfaStates,
                          sizeOfSyntaxDfaStates)};
}

void PersistentData::deleteSyntaxDfaStates(SyntaxDfaState **syntaxDfaStates,
                                           int sizeOfSyntaxDfaStates) {
  for (int i = 0; i < sizeOfSyntaxDfaStates; i++) {
    delete syntaxDfaStates[i];
  }
  delete syntaxDfaStates;
}

GetGrammarByInputStreamResult PersistentData::getGrammarByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetGrammarByInputStreamResult{false, nullptr};
  }
  int indexOfGrammar = readIntResult.data;
  return GetGrammarByInputStreamResult{true, grammars[indexOfGrammar]};
}

GetTokenDfaByInputStreamResult PersistentData::getTokenDfaByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetTokenDfaByInputStreamResult{false, nullptr};
  }
  int sizeOfTokenDfaStates = readIntResult.data;
  auto **tokenDfaStates = new TokenDfaState *[sizeOfTokenDfaStates];
  for (int indexOfTokenDfaState = 0;
       indexOfTokenDfaState < sizeOfTokenDfaStates; indexOfTokenDfaState++) {
    tokenDfaStates[indexOfTokenDfaState] = new TokenDfaState();
  }
  // countOfTokenDfaStates-(type-weight-terminal-countOfEdges-[ch,dest]{countOfEdges})
  for (int indexOfTokenDfaState = 0;
       indexOfTokenDfaState < sizeOfTokenDfaStates; indexOfTokenDfaState++) {
    TokenDfaState *tokenDfaState = tokenDfaStates[indexOfTokenDfaState];
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteTokenDfaStates(tokenDfaStates, sizeOfTokenDfaStates);
      return GetTokenDfaByInputStreamResult{false, nullptr};
    }
    tokenDfaState->type = readIntResult.data;
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteTokenDfaStates(tokenDfaStates, sizeOfTokenDfaStates);
      return GetTokenDfaByInputStreamResult{false, nullptr};
    }
    tokenDfaState->weight = readIntResult.data;
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteTokenDfaStates(tokenDfaStates, sizeOfTokenDfaStates);
      return GetTokenDfaByInputStreamResult{false, nullptr};
    }
    int intOfTerminal = readIntResult.data;
    if (intOfTerminal >= 0) {
      tokenDfaState->terminal = grammars[intOfTerminal];
    }
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteTokenDfaStates(tokenDfaStates, sizeOfTokenDfaStates);
      return GetTokenDfaByInputStreamResult{false, nullptr};
    }
    int sizeOfEdges = readIntResult.data;
    tokenDfaState->edges =
        std::unordered_map<byte, TokenDfaState *>(sizeOfEdges);
    for (int indexOfEdge = 0; indexOfEdge < sizeOfEdges; indexOfEdge++) {
      readIntResult = readInt();
      if (!readIntResult.isOk) {
        deleteTokenDfaStates(tokenDfaStates, sizeOfTokenDfaStates);
        return GetTokenDfaByInputStreamResult{false, nullptr};
      }
      int ch = readIntResult.data;
      readIntResult = readInt();
      if (!readIntResult.isOk) {
        deleteTokenDfaStates(tokenDfaStates, sizeOfTokenDfaStates);
        return GetTokenDfaByInputStreamResult{false, nullptr};
      }
      TokenDfaState *chToState = tokenDfaStates[readIntResult.data];
      std::pair<byte, TokenDfaState *> keyValue(ch, chToState);
      tokenDfaState->edges.insert(keyValue);
    }
  }

  return GetTokenDfaByInputStreamResult{
      true,
      new TokenDfa(tokenDfaStates[0], (const TokenDfaState **)tokenDfaStates,
                   sizeOfTokenDfaStates)};
}

void PersistentData::deleteTokenDfaStates(TokenDfaState **tokenDfaStates,
                                          int sizeOfTokenDfaStates) {
  for (int i = 0; i < sizeOfTokenDfaStates; i++) {
    delete tokenDfaStates[i];
  }
  delete tokenDfaStates;
}

GetDerivedTerminalGrammarAutomataDataByInputStreamResult
PersistentData::getDerivedTerminalGrammarAutomataDataByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetDerivedTerminalGrammarAutomataDataByInputStreamResult{false,
                                                                    nullptr};
  }
  auto *derivedTerminalGrammarAutomataData =
      new DerivedTerminalGrammarAutomataData();
  derivedTerminalGrammarAutomataData->count = readIntResult.data;

  if (derivedTerminalGrammarAutomataData->count == 0) {
    return GetDerivedTerminalGrammarAutomataDataByInputStreamResult{
        true, derivedTerminalGrammarAutomataData};
  }

  for (int indexOfRootTerminalGrammarMap = 0;
       indexOfRootTerminalGrammarMap <
       derivedTerminalGrammarAutomataData->count;
       indexOfRootTerminalGrammarMap++) {
    auto result = getRootTerminalGrammarMapByInputStream(
        &derivedTerminalGrammarAutomataData->rootTerminalGrammarMaps);
    if (!result.isOk) {
      delete derivedTerminalGrammarAutomataData;
      return GetDerivedTerminalGrammarAutomataDataByInputStreamResult{false,
                                                                      nullptr};
    }
  }

  return GetDerivedTerminalGrammarAutomataDataByInputStreamResult{
      true, derivedTerminalGrammarAutomataData};
}

GetRootTerminalGrammarMapByInputStreamResult
PersistentData::getRootTerminalGrammarMapByInputStream(
    std::vector<RootTerminalGrammarMap> *rootTerminalGrammarMaps) {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetRootTerminalGrammarMapByInputStreamResult{false};
  }

  Grammar *rootTerminalGrammar = grammars[readIntResult.data];

  readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetRootTerminalGrammarMapByInputStreamResult{false};
  }
  int terminalGrammarSize = readIntResult.data;

  auto textTerminalMap =
      std::unordered_map<std::string *, Grammar *, TextTerminalMapHash,
                         TextTerminalMapEq>(terminalGrammarSize);
  for (int indexOfKeyWords = 0; indexOfKeyWords < terminalGrammarSize;
       indexOfKeyWords++) {
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      return GetRootTerminalGrammarMapByInputStreamResult{false};
    }
    int intOfText = readIntResult.data;
    std::string *text = stringPool[intOfText];

    readIntResult = readInt();
    if (!readIntResult.isOk) {
      return GetRootTerminalGrammarMapByInputStreamResult{false};
    }
    int intOfTerminal = readIntResult.data;
    Grammar *terminal = grammars[intOfTerminal];

    std::pair<std::string *, Grammar *> pair(text, terminal);
    textTerminalMap.insert(pair);
  }
  rootTerminalGrammarMaps->push_back(
      RootTerminalGrammarMap(rootTerminalGrammar, textTerminalMap));
  return GetRootTerminalGrammarMapByInputStreamResult{true};
}

GetGrammarsByInputStreamResult PersistentData::getGrammarsByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetGrammarsByInputStreamResult{false, nullptr};
  }
  int _sizeOfGramamrs = readIntResult.data;
  auto **heapGrammars = new Grammar *[_sizeOfGramamrs];

  for (int indexOfGrammar = 0; indexOfGrammar < _sizeOfGramamrs;
       indexOfGrammar++) {
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteGrammars(heapGrammars,indexOfGrammar);
      return GetGrammarsByInputStreamResult{false, nullptr};
    }
    auto type = GrammarType(readIntResult.data);
    auto *grammar = newGrammarByType(type, indexOfGrammar);
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteGrammars(heapGrammars,indexOfGrammar);
      delete grammar;
      return GetGrammarsByInputStreamResult{false, nullptr};
    }
    grammar->name = *(stringPool[readIntResult.data]);
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteGrammars(heapGrammars,indexOfGrammar);
      delete grammar;
      return GetGrammarsByInputStreamResult{false, nullptr};
    }
    grammar->action = GrammarAction(readIntResult.data);
    if (type == GrammarType::TERMINAL) {
      auto *terminalGrammar = (TerminalGrammar *)grammar;
      readIntResult = readInt();
      if (!readIntResult.isOk) {
        deleteGrammars(heapGrammars,indexOfGrammar);
        delete grammar;
        return GetGrammarsByInputStreamResult{false, nullptr};
      }
      terminalGrammar->lookaheadMatchingMode =
          LookaheadMatchingMode(readIntResult.data);
    }
    heapGrammars[indexOfGrammar] = grammar;
  }
  this->sizeOfGramamrs = _sizeOfGramamrs;
  this->grammars = heapGrammars;
  return GetGrammarsByInputStreamResult{true, heapGrammars, _sizeOfGramamrs};
}

void PersistentData::deleteGrammars(Grammar **grammars, int sizeOfGrammars) {
  for(int i = 0 ; i < sizeOfGrammars;i++){
    delete grammars[i];
  }
  delete grammars;
}

Grammar *PersistentData::newGrammarByType(GrammarType type,
                                          int indexOfGrammar) {
  Grammar *grammar = nullptr;
  switch (type) {
  case GrammarType::TERMINAL:
    grammar = new TerminalGrammar(indexOfGrammar);
    break;
  case GrammarType::NONTERMINAL:
    grammar = new NonterminalGrammar(indexOfGrammar);
    break;
  case GrammarType::TERMINAL_FRAGMENT:
  default:
    break;
  }
  return grammar;
}

GetStringPoolByInputStreamResult PersistentData::getStringPoolByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetStringPoolByInputStreamResult{false, nullptr};
  }
  int sizeOfStrings = readIntResult.data;
  auto **strings = new std::string *[sizeOfStrings];
  for (int indexOfString = 0; indexOfString < sizeOfStrings; indexOfString++) {
    readIntResult = readInt();
    if (!readIntResult.isOk) {
      deleteStrings(strings,indexOfString);
      return GetStringPoolByInputStreamResult{false, nullptr};
    }
    int countOfStringBytes = readIntResult.data;
    ReadStringResult readStringResult = readByteString(countOfStringBytes);
    if (!readStringResult.isOk) {
      deleteStrings(strings,indexOfString);
      return GetStringPoolByInputStreamResult{false, nullptr};
    }
    std::string *str = readStringResult.data;
    strings[indexOfString] = str;
  }
  this->stringPool = strings;
  this->sizeOfStringPool = sizeOfStrings;
  return GetStringPoolByInputStreamResult{true, strings, sizeOfStrings};
}

void PersistentData::deleteStrings(std::string **strings, int sizeOfStrings) {
  for(int i = 0 ; i < sizeOfStrings;i++){
    delete strings[i];
  }
  delete strings;
}

ReadStringResult PersistentData::readByteString(int countOfStringBytes) {
  auto *str = new std::string(countOfStringBytes, 0);
  char *buf = const_cast<char *>(str->data());
  bool readResult = doRead((byte *)(buf), 0, countOfStringBytes);
  if(!readResult){
    delete str;
    return {false, nullptr};
  }
  return ReadStringResult{true, str};
}

ReadIntResult PersistentData::readInt() {
  bool readResult = doRead(intByteBuffer.buffer, 0, intByteBuffer.capacity);
  intByteBuffer.setPosition(intByteBuffer.capacity);
  return {readResult, intByteBuffer.getInt()};
}

bool PersistentData::doRead(byte bytes[], int offset, int length) {
  char *base = reinterpret_cast<char *>(bytes + offset);
  int countOfRead = inputStream.read(base, length).gcount();
  // 文件数据可能损坏
  return countOfRead == length;
}

BuildAutomataResult PersistentData::sourceDataError() {
  return {false, "data of automata File is error"};
}

void PersistentData::compact() { inputStream.close(); }

GetAstAutomataTypeResult PersistentData::getAstAutomataTypeByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetAstAutomataTypeResult{false};
  }
  auto type = AstAutomataType(readIntResult.data);
  return GetAstAutomataTypeResult{true, type};
}

GetNonterminalFollowMapByInputStreamResult
PersistentData::getNonterminalFollowMapByInputStream() {
  ReadIntResult readIntResult = readInt();
  if (!readIntResult.isOk) {
    return GetNonterminalFollowMapByInputStreamResult{false};
  }
  int size = readIntResult.data;
  auto *nonterminalFollowMap = new std::unordered_map<
      const Grammar *,
      std::unordered_set<const Grammar *, PtrGrammarContentHash,
                         PtrGrammarContentEq> *,
      PtrGrammarContentHash, PtrGrammarContentEq>(size);
  for (int indexOfNonterminal = 0; indexOfNonterminal < size;
       indexOfNonterminal++) {
    GetGrammarByInputStreamResult getGrammarByInputStreamResult =
        getGrammarByInputStream();
    if (!getGrammarByInputStreamResult.isOk) {
      delete nonterminalFollowMap;
      return GetNonterminalFollowMapByInputStreamResult{false};
    }
    Grammar *nonterminal = getGrammarByInputStreamResult.data;

    readIntResult = readInt();
    if (!readIntResult.isOk) {
      delete nonterminalFollowMap;
      return GetNonterminalFollowMapByInputStreamResult{false};
    }
    int sizeOfFollow = readIntResult.data;
    auto follow = new std::unordered_set<const Grammar *, PtrGrammarContentHash,
                                         PtrGrammarContentEq>(sizeOfFollow);
    for (int indexOfFollow = 0; indexOfFollow < sizeOfFollow; indexOfFollow++) {
      getGrammarByInputStreamResult = getGrammarByInputStream();
      if (!getGrammarByInputStreamResult.isOk) {
        delete nonterminalFollowMap;
        return GetNonterminalFollowMapByInputStreamResult{false};
      }
      follow->insert(getGrammarByInputStreamResult.data);
    }

    std::pair<const Grammar *,
              std::unordered_set<const Grammar *, PtrGrammarContentHash,
                                 PtrGrammarContentEq> *>
        pair(nonterminal, follow);
    nonterminalFollowMap->insert(pair);
  }
  return GetNonterminalFollowMapByInputStreamResult{true, nonterminalFollowMap};
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/PersistentData.java
package titan.ast.runtime;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * PersistentData对应在文件中的对象.
 *
 * @author tian wei jun
 */
public class PersistentData {
  InputStream byteInputStream;
  String[] stringPool;
  Grammar[] grammars;
  ProductionRule[] productionRules;

  ByteBuffer intByteBuffer = ByteBuffer.allocate(4);

  public PersistentData() {
    intByteBuffer.order(ByteOrder.BIG_ENDIAN);
  }

  public PersistentData(String persistentDataFilePath) {
    this();
    init(persistentDataFilePath);
  }

  public PersistentData(InputStream byteInputStream) {
    this();
    this.byteInputStream = byteInputStream;
  }

  private void init(String persistentDataFilePath) {
    try {
      this.byteInputStream = new FileInputStream(persistentDataFilePath);
    } catch (Exception e) {
      throw new AstRuntimeException(e);
    }
  }

  /**
   * 只能调用一次.
   *
   * @return
   */
  public String[] getStringPoolByInputStream() {
    int sizeOfStrings = readInt();
    String[] strings = new String[sizeOfStrings];
    for (int indexOfString = 0; indexOfString < sizeOfStrings; indexOfString++) {
      int countOfStringBytes = readInt();
      String str = readString(countOfStringBytes);
      strings[indexOfString] = str;
    }
    this.stringPool = strings;
    return strings;
  }

  /**
   * 只能调用一次.
   *
   * @return
   */
  public Grammar[] getGrammarsByInputStream() {
    int sizeOfGramamrs = readInt();
    Grammar[] grammars = new Grammar[sizeOfGramamrs];

    GrammarType[] grammarTypes = GrammarType.values();
    GrammarAction[] grammarActions = GrammarAction.values();
    LookaheadMatchingMode[] lookaheadMatchingModes = LookaheadMatchingMode.values();
    for (int indexOfGrammar = 0; indexOfGrammar < sizeOfGramamrs; indexOfGrammar++) {
      GrammarType type = grammarTypes[readInt()];
      Grammar grammar = generateGrammarByType(type, indexOfGrammar);
      grammar.name = stringPool[readInt()];
      grammar.action = grammarActions[readInt()];
      // lookaheadMatchingMode
      if (type == GrammarType.TERMINAL) {
        TerminalGrammar terminalGrammar = (TerminalGrammar) grammar;
        terminalGrammar.lookaheadMatchingMode = lookaheadMatchingModes[readInt()];
      }
      grammars[indexOfGrammar] = grammar;
    }
    this.grammars = grammars;
    return grammars;
  }

  private Grammar generateGrammarByType(GrammarType type, int indexOfGrammar) {
    Grammar grammar = null;
    switch (type) {
      case TERMINAL:
        grammar = new TerminalGrammar(indexOfGrammar);
        break;
      case NONTERMINAL:
        grammar = new NonterminalGrammar(indexOfGrammar);
        break;
      case TERMINAL_FRAGMENT:
      default:
    }
    return grammar;
  }

  /**
   * 只能调用一次.
   *
   * @return
   */
  public KeyWordAutomata getKeyWordAutomataByInputStream() {
    KeyWordAutomata keyWordAutomata = new KeyWordAutomata();

    keyWordAutomata.emptyOrNot = readInt();

    if (keyWordAutomata.emptyOrNot == KeyWordAutomata.EMPTY) {
      return keyWordAutomata;
    }

    keyWordAutomata.rootKeyWord = grammars[readInt()];

    int keyWordsSize = readInt();

    HashMap<String, Grammar> textTerminalMap = new HashMap<>(keyWordsSize);
    for (int indexOfKeyWords = 0; indexOfKeyWords < keyWordsSize; indexOfKeyWords++) {
      int intOfText = readInt();
      String text = stringPool[intOfText];

      int intOfTerminal = readInt();
      Grammar terminal = grammars[intOfTerminal];

      textTerminalMap.put(text, terminal);
    }
    keyWordAutomata.textTerminalMap = textTerminalMap;

    return keyWordAutomata;
  }

  /**
   * 只能调用一次.
   *
   * @return
   */
  public TokenDfa getTokenDfaByInputStream() {
    int sizeOfTokenDfaStates = readInt();
    TokenDfaState[] tokenDfaStates = new TokenDfaState[sizeOfTokenDfaStates];
    for (int indexOfTokenDfaState = 0;
        indexOfTokenDfaState < sizeOfTokenDfaStates;
        indexOfTokenDfaState++) {
      tokenDfaStates[indexOfTokenDfaState] = new TokenDfaState();
    }
    // countOfTokenDfaStates-(type-weight-terminal-countOfEdges-[ch,dest]{countOfEdges})
    for (int indexOfTokenDfaState = 0;
        indexOfTokenDfaState < sizeOfTokenDfaStates;
        indexOfTokenDfaState++) {
      TokenDfaState tokenDfaState = tokenDfaStates[indexOfTokenDfaState];
      tokenDfaState.type = readInt();
      tokenDfaState.weight = readInt();
      int intOfTerminal = readInt();
      if (intOfTerminal >= 0) {
        tokenDfaState.terminal = grammars[intOfTerminal];
      }
      int sizeOfEdges = readInt();
      tokenDfaState.edges = new HashMap<>(sizeOfEdges);
      for (int indexOfEdge = 0; indexOfEdge < sizeOfEdges; indexOfEdge++) {
        int ch = readInt();
        TokenDfaState chToState = tokenDfaStates[readInt()];
        tokenDfaState.edges.put(ch, chToState);
      }
    }
    TokenDfa tokenDfa = new TokenDfa();
    tokenDfa.start = tokenDfaStates[0];
    return tokenDfa;
  }

  public void getProductionRulesByInputStream() {
    int countOfProductionRules = readInt();
    ProductionRule[] productionRules = new ProductionRule[countOfProductionRules];
    for (int indexOfProductionRule = 0;
        indexOfProductionRule < countOfProductionRules;
        indexOfProductionRule++) {
      productionRules[indexOfProductionRule] = new ProductionRule();
    }
    this.productionRules = productionRules;

    for (ProductionRule productionRule : productionRules) {
      productionRule.grammar = grammars[readInt()];
      int indexOfAliasInStringPool = readInt();
      if (indexOfAliasInStringPool >= 0) {
        productionRule.alias = stringPool[indexOfAliasInStringPool];
      }
      productionRule.reducingDfa = getSyntaxDfaByInputStream();
    }
  }

  public SyntaxDfa getSyntaxDfaByInputStream() {
    int sizeOfSyntaxDfaStates = readInt();
    SyntaxDfaState[] syntaxDfaStates = new SyntaxDfaState[sizeOfSyntaxDfaStates];
    for (int indexOfSyntaxDfaState = 0;
        indexOfSyntaxDfaState < sizeOfSyntaxDfaStates;
        indexOfSyntaxDfaState++) {
      syntaxDfaStates[indexOfSyntaxDfaState] = new SyntaxDfaState(indexOfSyntaxDfaState);
    }
    // countOfSyntaxDfaStates-(type-countOfEdges-[ch,dest]{countOfEdges}-countOfProductions-productions)
    for (int indexOfSyntaxDfaState = 0;
        indexOfSyntaxDfaState < sizeOfSyntaxDfaStates;
        indexOfSyntaxDfaState++) {
      SyntaxDfaState syntaxDfaState = syntaxDfaStates[indexOfSyntaxDfaState];
      syntaxDfaState.type = readInt();
      int sizeOfEdges = readInt();
      syntaxDfaState.edges = new HashMap<>(sizeOfEdges);
      for (int indexOfEdge = 0; indexOfEdge < sizeOfEdges; indexOfEdge++) {
        Grammar ch = grammars[readInt()];
        SyntaxDfaState chToState = syntaxDfaStates[readInt()];
        syntaxDfaState.edges.put(ch, chToState);
      }
      int sizeOfProductions = readInt();
      syntaxDfaState.closingProductionRules = new ArrayList<>(sizeOfProductions);
      for (int indexOfProduction = 0; indexOfProduction < sizeOfProductions; indexOfProduction++) {
        syntaxDfaState.closingProductionRules.add(productionRules[readInt()]);
      }
    }
    SyntaxDfa syntaxDfa = new SyntaxDfa();
    syntaxDfa.start = syntaxDfaStates[0];
    return syntaxDfa;
  }

  private String readString(int countOfStringBytes) {
    byte[] bytes = new byte[countOfStringBytes];
    doRead(bytes, 0, countOfStringBytes);

    char[] chars = new char[countOfStringBytes];
    int indexOfChar = 0;
    while (indexOfChar < countOfStringBytes) {
      byte b = bytes[indexOfChar];
      chars[indexOfChar] = (char) b;
      ++indexOfChar;
    }
    return new String(chars);
  }

  private int readInt() {
    doRead(intByteBuffer.array(), 0, intByteBuffer.capacity());
    intByteBuffer.clear();
    intByteBuffer.limit(intByteBuffer.capacity());
    return intByteBuffer.getInt();
  }

  private void doRead(byte[] bytes, int offset, int length) {
    try {
      int countOfRead = byteInputStream.read(bytes, offset, length);
      if (countOfRead != length) {
        throw new AstRuntimeException("countOfRead is not equal the expectant number.");
      }
    } catch (IOException e) {
      closeInputStream();
      throw new AstRuntimeException(e);
    }
  }

  public void closeInputStream() {
    if (null != byteInputStream) {
      try {
        byteInputStream.close();
      } catch (IOException e) {
        throw new AstRuntimeException(e);
      }
    }
  }

  public void compact() {
    closeInputStream();
    intByteBuffer = null;
  }

  public AstAutomataType getAstAutomataTypeByInputStream() {
    int original = readInt();
    return AstAutomataType.values()[original];
  }

  public Map<Grammar, Set<Grammar>> getNonterminalFollowMapByInputStream() {
    int size = readInt();
    Map<Grammar, Set<Grammar>> nonterminalFollowMap = new HashMap<>(size);
    for (int indexOfNonterminal = 0; indexOfNonterminal < size; indexOfNonterminal++) {
      Grammar nonterminal = getGrammarByInputStream();
      int sizeOfFollow = readInt();
      Set<Grammar> follow = new HashSet<>(sizeOfFollow);
      for (int indexOfFollow = 0; indexOfFollow < sizeOfFollow; indexOfFollow++) {
        follow.add(getGrammarByInputStream());
      }
      nonterminalFollowMap.put(nonterminal, follow);
    }

    return nonterminalFollowMap;
  }

  public Grammar getGrammarByInputStream() {
    int indexOfGrammar = readInt();
    return grammars[indexOfGrammar];
  }
}
